### ISSUE TITLE
4.0.1 regression

### ISSUE THREAD

**Issue by oliviertassinari** [mozilla/node-convict#224]
It seems that the `default` is no longer respected as soon as we use an `env` key and this value is missing. I haven't looked to much into the issue. Maybe someone else will. I'm reverting to `4.0.0`.

**Comment by madarche** [comment]
@oliviertassinari can you provide an example please?

**Comment by oliviertassinari** [comment]
@madarche For instance, the `convict.get('session.secret')` isn't present with `v4.0.1` but is with `v4.0.0`.

```js
session: {
  secret: {
    default: 'bubu',
    env: 'SESSION_SECRET',
    format: String,
  },
},
```

**Comment by kevinoneill** [comment]
I think this is to do with #217 which changes the behaviour of empty string from triggering the default to supplying the empty string.

In my case, the `4.0.0` ->  `4.0.1` update was a generated as part of a `yarn install` that updated convict as a side effect. 

A change like this (even if it represents a bug fix) should have been flagged as breaking.

**Comment by oliviertassinari** [comment]
@kevinoneill I think that you are right. I'm running node-convict in a docker environment. I have the following warning: `The SESSION_SECRET variable is not set. Defaulting to a blank string.`. With the new behavior of the project. I do no longer see the default value set.

**Comment by madarche** [comment]
A PR with the corresponding test is welcome.

**Comment by mojavelinux** [comment]
If the environment variable is present and the value of the variable is empty string, the config value should be set to empty string. If the environment variable is missing, the default value should be used. Is that the correct assumption?

**Comment by kevinoneill** [comment]
Seems right to me.

**Comment by salper** [comment]
Well... That is the regression. Previously, a `FOO=` would use the default value. Since the patch, it overrides the default value. Some services which rely on declaring statically environment variables like docker-compose are impacted by this kind of changes.

**Comment by mojavelinux** [comment]
I agree this change warranted a major version. At this point, though, there are probably users relying on both behaviors, so that's water over the damn. What we need to decide is how to move forward.

If the decision is made to rollback, there still needs to be a way to unset a config value using an environment variable as the environment variable is the last resort to override a value. One option is to recognize the special value "undefined" as an intent to unset the value. Another option is to use a companion environment variable whose sole purpose is to unset the value (e.g., NO_SESSION_SECRET=true). (This is similar to a negated CLI option like --no-session-secret). If anyone has other ideas, please propose them.

**Comment by A-312** [comment]
> If anyone has other ideas, please propose them.

CustomGetter #313 people rewrite env and arg function like they want.

**Comment by A-312** [comment]
```js
convict.addGetter('env', function(value, schema, stopPropagation) {
  if (value in this.getEnv()) {
    stopPropagation();
    return schema._cvtCoerce(this.getEnv()[value]);
  }
}, false, true);
```
